go-controller/pkg/network-controller-manager/network_controller_manager.go

```diff
@@ -0,0 +1,282 @@
+	if err != nil {
+		return err
+	}
+
+	cm.wg.Add(1)
+	go func() {
+		leaderElector.Run(ctx)
+		klog.Infof("Stopped leader election")
+		cm.wg.Done()
+	}()
+
+	return nil
+}
+
+// NewNetworkControllerManager creates a new OVN controller manager to manage all the controller for all networks
+func NewNetworkControllerManager(ovnClient *util.OVNClientset, identity string, wf *factory.WatchFactory,
+	libovsdbOvnNBClient libovsdbclient.Client, libovsdbOvnSBClient libovsdbclient.Client,
+	recorder record.EventRecorder, wg *sync.WaitGroup) *NetworkControllerManager {
+	podRecorder := metrics.NewPodRecorder()
+
+	return &NetworkControllerManager{
+		client: ovnClient.KubeClient,
+		kube: &kube.Kube{
+			KClient:              ovnClient.KubeClient,
+			EIPClient:            ovnClient.EgressIPClient,
+			EgressFirewallClient: ovnClient.EgressFirewallClient,
+			CloudNetworkClient:   ovnClient.CloudNetworkClient,
+		},
+		stopChan:     make(chan struct{}),
+		watchFactory: wf,
+		recorder:     recorder,
+		nbClient:     libovsdbOvnNBClient,
+		sbClient:     libovsdbOvnSBClient,
+		podRecorder:  &podRecorder,
+
+		wg:             wg,
+		identity:       identity,
+		ovnControllers: make(map[string]NetworkController),
+	}
+}
+
+// Init initializes the controller manager and create/start default controller
+func (cm *NetworkControllerManager) Init() error {
+	cm.configureMetrics(cm.stopChan)
+
+	err := cm.configureSCTPSupport()
+	if err != nil {
+		return err
+	}
+
+	err = cm.enableOVNLogicalDataPathGroups()
+	if err != nil {
+		return err
+	}
+
+	cm.NewDefaultNetworkController()
+	return nil
+}
+
+func (cm *NetworkControllerManager) configureSCTPSupport() error {
+	hasSCTPSupport, err := util.DetectSCTPSupport()
+	if err != nil {
+		return err
+	}
+
+	if !hasSCTPSupport {
+		klog.Warningf("SCTP unsupported by this version of OVN. Kubernetes service creation with SCTP will not work ")
+	} else {
+		klog.Info("SCTP support detected in OVN")
+	}
+	cm.SCTPSupport = hasSCTPSupport
+	return nil
+}
+
+// enableOVNLogicalDataPathGroups sets an OVN flag to enable logical datapath
+// groups on OVN 20.12 and later. The option is ignored if OVN doesn't
+// understand it. Logical datapath groups reduce the size of the southbound
+// database in large clusters. ovn-controllers should be upgraded to a version
+// that supports them before the option is turned on by the master.
+func (cm *NetworkControllerManager) enableOVNLogicalDataPathGroups() error {
+	nbGlobal := nbdb.NBGlobal{
+		Options: map[string]string{"use_logical_dp_groups": "true"},
+	}
+	if err := libovsdbops.UpdateNBGlobalSetOptions(cm.nbClient, &nbGlobal); err != nil {
+		return fmt.Errorf("failed to set NB global option to enable logical datapath groups: %v", err)
+	}
+	return nil
+}
+
+func (cm *NetworkControllerManager) configureMetrics(stopChan <-chan struct{}) {
+	metrics.RegisterMasterPerformance(cm.nbClient)
+	metrics.RegisterMasterFunctional()
+	metrics.RunTimestamp(stopChan, cm.sbClient, cm.nbClient)
```

Comment:
Hmm I missed that `RunTimestamp` needs to preferably be in the `Run` method. But I am not sure if we need to do this before we write to the database. I will check with @martinkennelly. Should not be an actual issue because Init and Run are called in sequence so lets go with things like they are now.